File: python/23_MergekSortedLists.py
from unittest import TestCase
import queue
from Definitions import ListNode
# https://leetcode.com/problems/merge-k-sorted-lists/


class MergekSortedLists(TestCase):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        tail = head = ListNode(0)
        que = queue.PriorityQueue()
        for i in range(len(lists)):
            lists[i] and que.put((lists[i].val, i))
        while not que.empty():
            i = que.get()[1]
            tail.next = lists[i]
            tail = tail.next
            lists[i] = lists[i].next
            lists[i] and que.put((lists[i].val, i))
        return head.next

    def test1(self):
        self.assertEqual([1, 1, 2, 3, 4, 4, 5, 6], self.mergeKLists([
            ListNode.build_from_list(1, 4, 5),
            ListNode.build_from_list(1, 3, 4),
            ListNode.build_from_list(2, 6),
        ]).as_list())

File: kotlin/src/main/kotlin/23.MergekSortedLists.kt
import ListNode.Companion.buildFromList
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

//https://leetcode.com/problems/merge-k-sorted-lists/

class MergekSortedLists {
  fun mergeKLists(lists: Array<ListNode?>): ListNode? {
    val que = PriorityQueue<ListNode>{x: ListNode, y:ListNode -> x.`val` - y.`val`}

    for(node in lists)
      if(node != null) que.add(node)

    val head = ListNode(0)
    var tail: ListNode = head
    while(!que.isEmpty()){
      var node = que.poll()
      tail.next = node
      node = node.next
      tail = tail.next!!
      if(node != null) que.add(node)
    }

    return head.next
  }

  @Test
  fun test1(){
    assertEquals(listOf(1,1,2,3,4,4,5,6), mergeKLists(arrayOf(
        buildFromList(1,4,5),
        buildFromList(1,3,4),
        buildFromList(2,6)
    ))!!.asList())
  }
}